File: server/src/routes/images.ts
/**
 * Image routes - Image generation and inpainting endpoints
 * 
 * Provides:
 * - POST /api/images/generate - Image generation/editing
 * - POST /api/images/inpaint - DEPRECATED (migrated to Python server)
 */

import { Router, type Request, type Response } from 'express';
import { createGeminiService } from '../services/geminiService.js';
import { createImageGenerationService } from '../services/imageGenerationService.js';
import { asyncHandler, APIError } from '../middleware/errorHandler.js';
import { requireGeminiApiKey, type ApiKeyRequest } from '../middleware/apiKeyValidation.js';
import type {
  GenerateImageResponse,
  GeminiRawResponse,
} from '../types/api.js';
import { generateImageRequestSchema } from '../schemas/index.js';

const router = Router();

/**
 * POST /api/images/generate
 * 
 * Image generation/editing endpoint using Gemini
 * 
 * Request body: GenerateImageRequest
 * Response: GenerateImageResponse
 */
router.post('/generate', requireGeminiApiKey, asyncHandler(async (req: Request, res: Response) => {
  // Validate request body with Zod
  const {
    model,
    sourceImage,
    prompt,
    logLabel: _logLabel,
  } = generateImageRequestSchema.parse(req.body);

  // Create services (API key validated by middleware)
  const gemini = createGeminiService((req as ApiKeyRequest).geminiApiKey);
  const imageService = createImageGenerationService(gemini);

  try {
    const result = await imageService.generateImage({
      sourceImage,
      prompt,
      model,
    });

    const response: GenerateImageResponse = {
      raw: result.raw as GeminiRawResponse,
      imageData: result.imageData,
    };

    res.json(response);

  } catch (error) {
    // Handle Gemini API errors
    if (error instanceof Error) {
      throw new APIError(
        500,
        'Image generation failed',
        error.message
      );
    }
    throw error;
  }
}));

/**
 * POST /api/images/inpaint
 * 
 * DEPRECATED: This endpoint has been migrated to the Python server.
 * Use /api/images/inpaint on the Python server (port 8001) instead.
 */
router.post('/inpaint', asyncHandler(async (_req: Request, _res: Response) => {
  throw new APIError(501, 'DEPRECATED: Use Python server at /api/images/inpaint instead');
}));

export default router;

File: python-server/schemas/images.py
"""
Pydantic schemas for image generation and inpainting endpoints.

These schemas match the TypeScript interfaces in server/src/types/api.ts
to ensure API compatibility between Express and Python backends.
"""

from typing import Optional

from pydantic import BaseModel, Field

from .agentic import Base64ImageUrl


# =============================================================================
# POST /api/images/generate - Image generation/editing
# =============================================================================


class GenerateImageRequest(BaseModel):
    """
    Request body for POST /api/images/generate endpoint.

    Matches the TypeScript GenerateImageRequest interface and Zod schema.
    """

    model: str = Field(..., min_length=1, description="The model to use")
    sourceImage: Base64ImageUrl = Field(
        ..., description="Source image as base64 data URL"
    )
    prompt: str = Field(..., min_length=1, description="Edit prompt")
    maskImage: Optional[Base64ImageUrl] = Field(
        None, description="Optional mask image as base64 data URL"
    )
    isImageGeneration: Optional[bool] = Field(
        None, description="Whether this is an image generation call"
    )
    logLabel: Optional[str] = Field(None, description="Label for this call in the log")


class GenerateImageResponse(BaseModel):
    """
    Response for POST /api/images/generate endpoint.

    Matches the TypeScript GenerateImageResponse interface.
    """

    raw: dict = Field(..., description="The raw result from the API")
    imageData: str = Field(..., description="Generated image as base64 data URL")


# =============================================================================
# POST /api/images/inpaint - Two-step inpainting
# =============================================================================


class InpaintRequest(BaseModel):
    """
    Request body for POST /api/images/inpaint endpoint.

    Matches the TypeScript InpaintRequest interface and Zod schema.
    """

    sourceImage: Base64ImageUrl = Field(
        ..., description="Source image as base64 data URL"
    )
    maskImage: Base64ImageUrl = Field(..., description="Mask image as base64 data URL")
    prompt: str = Field(..., min_length=1, description="Edit prompt")
    thinkingBudget: Optional[int] = Field(
        None, description="Thinking budget for planning"
    )


class InpaintResponse(BaseModel):
    """
    Response for POST /api/images/inpaint endpoint.

    Matches the TypeScript InpaintResponse interface.
    """

    imageData: str = Field(..., description="Generated image as base64 data URL")
    refinedPrompt: str = Field(
        ..., description="AI's refined prompt used for generation"
    )
    thinking: str = Field(..., description="AI's thinking during planning")
